### PR TITLE
refactor(components): reduce the amount of reduce we do in DataList

### PR DESCRIPTION
**Pull request by darryltec**
## Motivations

This started off as fixing that issue with the layout not changing when the screen changes from wide to small (or vice-versa) but it went more than that now. It ended up removing the need to use a reducer and coupling of layout, header, and loading state. Which in turn made (IMO) the code more readable, organized, and reactive to change with fewer maps and reducers.

This also changed how we deal with the loading state on mobile and closely matches the layout now instead of using a glimmer text.

### Added

- Registering layouts to the context
- Hooks for determining which layout to use
- Separate components for list items, list item, and headers

### Changed

- How we handle determining which layout to use

### Deprecated

- 

### Removed

- unused components

### Fixed

- 

### Security

- 

## Testing

Aside from the loading state on smaller screens, there should be no change to how it should show up on the user (sans bugs).

---

[In Atlantis we use Github's built in pull request reviews](https://help.github.com/en/articles/about-pull-request-reviews).

Random photo of Atlantis

### CODE REVIEW

File: packages/components/src/DataList/hooks/useActiveLayout.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import { useResponsiveSizing } from "./useResponsiveSizing";
+import { useDataListContext } from "../context/DataListContext";
+
+export function useActiveLayout() {
+  const { layoutBreakpoints, layouts } = useDataListContext();
+  const breakpoints = useResponsiveSizing();
+  const reversedLayoutBreakpoints = [...layoutBreakpoints].reverse();
```
Comment (rodrigoeidelvein) [inline packages/components/src/DataList/hooks/useActiveLayout.ts:7]: Why did you revert it?

Comment (darryltec) [reply packages/components/src/DataList/hooks/useActiveLayout.ts:7]: So it checks the highest breakpoint first and see if it's true

File: packages/components/src/DataList/components/DataListLayoutActions/DataListLayoutActions.tsx [hunk 2]
```diff
@@ -25,6 +26,10 @@ export function DataListLayoutActions() {
     };
   }, []);
 
+  if (loadingState === "initial") {
+    return <Glimmer shape="square" size="large" />;
+  }
+
```
Comment (rodrigoeidelvein) [inline packages/components/src/DataList/components/DataListLayoutActions/DataListLayoutActions.tsx:32]: Loved it!

File: packages/components/src/DataList/DataList.tsx [hunk 3]
```diff
@@ -132,33 +151,14 @@ function InternalDataList() {
...
-      {initialLoading && (
-        <DataListLoadingState
-          headers={headers}
-          layouts={layoutComponents}
-          mediaMatches={mediaMatches}
-        />
-      )}
+      {initialLoading && <DataListLoadingState />}
 
       {showEmptyState && <InternalDataListEmptyState />}
```
Comment (rodrigoeidelvein) [inline packages/components/src/DataList/DataList.tsx:159]: An improvement that could be made here is to check if we should render these components inside the component itself, as we're doing for DataListHeader, and others.

But, I can also see the point of being more readable this way.

**Review by rodrigoeidelvein — APPROVED** [review]
Just a minor comment with a suggestion that might don't make sense actually haha just food for thought

Looks really good, cleaner and works the same way! Have you done any performance test to see if removing the hook with the reduces changed anything?

File: packages/components/src/DataList/DataList.tsx (hunk 3 above)
Comment (darryltec) [reply packages/components/src/DataList/DataList.tsx:159]: Hmmm that's definitely something to think about!

Comment (darryltec) [reply packages/components/src/DataList/DataList.tsx:159]: I'll merge this first tho

**Comment by darryltec** [comment]
> Have you done any performance test to see if removing the hook with the reduces changed anything?

Yup. mostly that i couldn't replicate the issue with the layout not changing anymore compared to before.

**Pull request merged**